Add support for litex_vexriscv GPIO interrupts

robertszczepanski opened the pull request:
This adds support for GPIO interrupts which has been added to LiteX recently: https://github.com/enjoy-digital/litex/blob/master/litex/soc/cores/gpio.py#L21

Commit:
drivers: gpio_litex: add support for litex_vexriscv GPIO interrupts

This commit adds support for GPIO interrupts in GPIO driver for Litex
SoC Builder.

Signed-off-by: Robert Szczepanski <rszczepanski@internships.antmicro.com>

mnkp commented on drivers/gpio/Kconfig.litex:
```diff
@@ -8,3 +8,9 @@ config GPIO_LITEX
 	help
 	  Enable Litex GPIO driver.
+
+config GPIO_LITEX_IRQ
```
On other SoCs we generally don't have Kconfig options to enable IRQ support in GPIO driver. The savings are minimal. Also, the option is easy to overlook by the user. Maybe we should remove it? I.e. the IRQ support should be always enabled.

robertszczepanski replied:
I did get rid of that option so GPIO IRQs are always enabled right now.

mnkp commented on drivers/gpio/gpio_litex.c:
```diff
@@ -191,16 +189,45 @@ static int gpio_litex_port_toggle_bits(const struct device *dev,
+	const struct gpio_litex_cfg *gpio_config = DEV_GPIO_CFG(dev);
+	int ret = -1;
+
+	if (mode == GPIO_INT_MODE_EDGE) {
+		uint8_t ev_enabled = litex_read(GPIO_EV_ENABLE(dev),
+				gpio_config->reg_size);
+		litex_write(GPIO_EV_ENABLE(dev), gpio_config->reg_size,
+			    ev_enabled | BIT(pin));
+		ret = 0;
 	}
```
Sorry, I missed it before. It looks like the driver supports edge interrupts on both edges only. We should return `-ENOTSUP` in case user tries to configure level interrupt or edge interrupt on a single edge.

It would be good to test the driver with `tests/drivers/gpio/gpio_basic_api/` test.

robertszczepanski replied:
I changed `int ret = -1` to `int ret= -ENOTSUP` so there will be `-ENOTSUP` returned always if user tries to configure interrupt different than edge interrupt. Let me know if that's okay.

I don't think I am able to test this with `gpio_basic_api` since it requires to have `in-gpios` and `out-gpios` on the same GPIO port. I am using Arty A7 development board with LiteX SoC Builder. Unfortunately LiteX does not support interrupts on `GPIOInOut`. It is supported on `GPIOIn` only. That's what I was testing - multiple interrupts on multiple `GPIOIn` modules in `GPIO_INT_EDGE_RISING` or `GPIO_INT_EDGE_FALLING` modes. 
I've also tested it with `GPIO_INT_EDGE_BOTH` mode but since it is not supported by LiteX it works just like falling/rising edge mode depending on IRQ setup in LiteX `GPIOIn` module.

robertszczepanski replied:
According to my previous coment I decided to add a condition `if(gpio_config->port_is_output == true)` which returns `-ENOTSUP` if port is an output port. 
I did get rid of `int ret` as well so I think it is more clear what function returns now.

madscientist159 commented on drivers/gpio/gpio_litex.c:
```diff
@@ -38,13 +40,22 @@ struct gpio_litex_cfg {
+#define GPIO_EV_PENDING(dev) \
+		((volatile uint32_t *)((uint32_t)DEV_GPIO_CFG(dev)->reg_addr + 0x0008))
+#define GPIO_EV_ENABLE(dev) \
+		((volatile uint32_t *)((uint32_t)DEV_GPIO_CFG(dev)->reg_addr + 0x000c))
```
Same issue as above -- our build uses uint8_t and 0x14 for this register.  Note the CSR is configurable in LiteX, so this type of issue will not be unique to our builds.

madscientist159 reviewed (changes requested):
This still needs some work to be mergeable, given its location in the tree as a general driver for the LiteX GPIO peripheral.

madscientist159 commented on drivers/gpio/gpio_litex.c:
```diff
@@ -191,17 +189,49 @@ static int gpio_litex_port_toggle_bits(const struct device *dev,
 	}
-	return ret;
+
+	return -ENOTSUP;
```
This function should provide a means to disable an active interrupt by passing `GPIO_INT_DISABLE` as the mode.

robertszczepanski replied:
Thanks for pointing it oout. I've added support for disabling interrupts in latest commit. It has been tested on Arty A7 with LiteX SoC and it works great.

mnkp replied:
@madscientist159 thanks for spotting.

Commit:
drivers: gpio_litex: make IRQ CSR addresses adjustable for user

This adds an option to set IRQ pending and IRQ enabled CSR adresses
in devicetree since these can be custom in LiteX.

Signed-off-by: Raptor Engineering Development Team <support@raptorengineering.com>

madscientist159 commented:
Here's a patch that resolves the device tree problems (but not the IRQ disable problems):
https://gitlab.raptorengineering.com/kestrel-collaboration/kestrel-firmware/zephyr-rtos/-/commit/4769452c23d02bd3635669205dd5a62261144b4c

And an example DT entry:
https://gitlab.raptorengineering.com/kestrel-collaboration/kestrel-firmware/zephyr-rtos/-/commit/34b7ae41364bc6666f86eb897602248e17f17945

Commit:
drivers: gpio_litex: add support for disabling IRQs and update devicetree

This adds missing option to disable IRQs.
Devicetree is modified to match previous commit with custom IRQ CSR
addresses.

Signed-off-by: Robert Szczepanski <rszczepanski@internships.antmicro.com>

robertszczepanski commented:
@madscientist159 since your solution looks nice for our custom CSR offset issue I cherry-picked this commit from the Kestrel Project: https://gitlab.raptorengineering.com/kestrel-collaboration/kestrel-firmware/zephyr-rtos/-/commit/4769452c23d02bd3635669205dd5a62261144b4c. However the DCO checker is not happy with the commit https://github.com/zephyrproject-rtos/zephyr/pull/33778/checks?check_run_id=2410239087. Are you OK with signing-off this commit? Just to be on the same page - signing-off the commit means you agree with https://docs.zephyrproject.org/latest/contribute/index.html#developer-certification-of-origin-dco.

madscientist159 commented:
@robertszczepanski Yes, that commit can be signed off by Raptor according to the linked terms.  If this authorization here is sufficient please feel free to add the line, otherwise I can push a modified commit to a different branch.

robertszczepanski commented:
@madscientist159 Great, thanks! I changed commit message and it passed CI check so it is good now.

madscientist159 reviewed (approved):
Looks good with the changes!

robertszczepanski commented:
@mnkp could you take a look at CI tests? For some reason it fails and it doesn't look like something relevant to this PR.
1: https://github.com/zephyrproject-rtos/zephyr/pull/33778/checks?check_run_id=2419806413
2: https://github.com/zephyrproject-rtos/zephyr/pull/33778/checks?check_run_id=2419806439

Commit:
drivers: gpio_litex: add support for changing IRQ type

LiteX GPIOIn module provides possibility to change IRQ mode
and edge via CSRs. This commit adds support for that feature.

Signed-off-by: Robert Szczepanski <rszczepanski@internships.antmicro.com>

robertszczepanski commented:
I noticed that when I created this PR LiteX has already added support for changing interrupts mode and edge by CSRs. I've added commit that provides support for this feature in Zephyr driver. I tested it again on Arty A7 and it works as expected.

mnkp commented:
There have been quite a few errors in the last CI run but at least some of them are legitimate errors that have already been fixed on master. Let's rebase this PR on the latest master and see what happens.

robertszczepanski commented:
I resolved latest conflict with 1b5542056b818e224fc86fb33862bf46c2277707. 

I would appreciate if someone looked at this and got it merged.

kgugala reviewed (approved):
LGTM
